### DISCUSSION TITLE
How to get specific source code before buildModule hook

### DISCUSSION THREAD

**Discussion by TomokiMiyauci** [webpack/webpack#15475] [Q&A]
I am creating a webpack-plugin.

There is a scenario where a CSS file is dynamically created based on a specific source code (e.g., .tsx). The creation of the CSS file is done once with all source code combined. This means that sequential execution is not possible for each source code fetch.

The CSS file is then written out as a virtual module.

Is there a good way to accomplish this scenario?

In The `buildModule` hook, `RawSource` does not exist. (
I don't want to read file)
The `succeedModule` hook has a `RawSource`, but it is already built, so updates to the virtual module will not be reflected.

---

Reference

The example is an entry point, importing a virtual module.

```index.ts
// index.ts
import "map.css"
import A from "A.tsx"
import B from "B.tsx"
````

```A.tsx
// A.tsx
export default A = () => <div className="block">A</div>
```

```B.tsx
// B.tsx
export default B = () => <div className="inline-block">B</div>
```

The virtual module `map.css` is output with the classes extracted from the `.tsx` file mapped as styles.

```map.css
.block{display:block}.inline-block{display:inline-block}
```

[mapcss](https://github.com/TomokiMiyauci/mapcss) maps classes to CSS Statements.

**Comment by vankop** [comment]
not sure what you expect.. probably 
```js
import "map.css"
import A from "A.tsx"
import B from "B.tsx"
```
is wrong idea, since `map.css` depends on `a.tsx` and `b.tsx`. Maybe you just need some extracted asset in the end of compilation?
So you will have loader + plugin. 
loader.js
```js
const infoByModule = new WeakMap();
module.exports = function (content) {
 const info = processContent(content);
 infoByModule.set(this._module, info);
 return content;
}
```
plugin.js
```js
class plugin {
   apply(compiler) {
      compiler.hooks.compilation.tap("plugin", compilation => {
          let css = "";
           compilation.hooks.finishModules.tapAsync("plugin", modules => {
                for (const module of modules) {
                    const info = infoByModule.get(module);
                    if (!info) continue;
                    css = addCss(css, info);
                }
           });
           	compilation.hooks.processAssets.tap(
		{
			name: "plugin",
			stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONS
		},
		() => {
		   compilation.emitAsset("name.css", new RawSource(css), {});
                });	
      });
   }
}
```

**Reply by TomokiMiyauci** [reply]
Thanks for the thorough explanation!

Now, the stylesheet is output as an asset.
https://github.com/TomokiMiyauci/mapcss/blob/beta/webpack/mod.js

I am not explaining it well enough, 
but I am assuming that the output CSS will eventually be inlined in the HTML head tag, or used as CSS Modules.

In other words, I intended to be able to use it as CSS Modules with `css-loader` or to output it inline with `style-loader`.

Therefore, I will need the following import statement.

```ts
import "map.css"
```

or

```ts
import { block } from "map.css"
```

"map.css" does not actually exist, but it scans content and generates virtual modules before the `css-loader`, etc.

Any opinions on this approach?

**Reply by vankop** [reply]
no, this approach will not work. webpack construct module graph from source modules. In your case you want to make module dependent on result of compilation.. so there is no way to do this since compilation depends on module graph..

**Reply by TomokiMiyauci** [reply]
I know exactly what you mean.

In fact, there is no need to refer to the module graph, I can refer to the file directly.
In the end, it is important for me when and how to acquire the content in question.

I have at least two options.

- ReadFile all the target content myself at some point, such as immediately after `apply`; at HMR, readFile only the updated files.
- Prepare a placeholder in the module and do string substitution afterwards.

The latter is the approach being taken with [this project](https://github.com/unocss/unocss/blob/main/packages/webpack/src/index.ts#L61), but it seems pretty hacky.

The former seems to have a performance impact, but is this a bad practice?

**Comment by BUPTlhuanyu** [comment]
I face the same problem...
